File: features/serialization_context/sercontext/SerContext.java
package serialization_context.sercontext;

import com.google.protobuf.ByteString;
import io.temporal.api.common.v1.Payload;
import io.temporal.api.common.v1.Payloads;
import io.temporal.api.history.v1.History;
import io.temporal.api.history.v1.HistoryEvent;
import io.temporal.common.converter.CodecDataConverter;
import io.temporal.common.converter.DataConverter;
import io.temporal.common.converter.DefaultDataConverter;
import io.temporal.payload.context.ActivitySerializationContext;
import io.temporal.payload.context.SerializationContext;
import io.temporal.payload.context.WorkflowSerializationContext;
import java.util.Collections;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Predicate;

/**
 * Context aware converters shared by the serialization_context features, plus history helpers used
 * to assert on the recorded context.
 */
public final class SerContext {

  public static final String METADATA_KEY = "ctx-signature";
  public static final String NO_CONTEXT = "none";
  public static final String DEFAULT_FAILURE_SOURCE = "JavaSDK";

  /**
   * Every signature the codec has been asked to encode or decode with. Worker and client share a
   * process here, so this is how a feature asserts on a context whose payload never shows up in
   * history.
   */
  public static final Set<String> OBSERVED_SIGNATURES = ConcurrentHashMap.newKeySet();

  private SerContext() {}

  public static String workflowSignature(String namespace, String workflowId) {
    return "wf|" + namespace + "|" + workflowId;
  }

  public static String activitySignature(
      String namespace,
      String workflowId,
      String workflowType,
      String activityType,
      String activityTaskQueue,
      boolean local) {
    return "act|"
        + namespace
        + "|"
        + workflowId
        + "|"
        + workflowType
        + "|"
        + activityType
        + "|"
        + activityTaskQueue
        + "|"
        + local;
  }

  public static String signature(SerializationContext context) {
    if (context instanceof WorkflowSerializationContext) {
      WorkflowSerializationContext workflow = (WorkflowSerializationContext) context;
      return workflowSignature(workflow.getNamespace(), workflow.getWorkflowId());
    }
    if (context instanceof ActivitySerializationContext) {
      ActivitySerializationContext activity = (ActivitySerializationContext) context;
      return activitySignature(
          activity.getNamespace(),
          activity.getWorkflowId(),
          activity.getWorkflowType(),
          activity.getActivityType(),
          activity.getActivityTaskQueue(),
          activity.isLocal());
    }
    return NO_CONTEXT;
  }

  public static DataConverter dataConverter() {
    return new CodecDataConverter(
        DefaultDataConverter.newDefaultInstance().withFailureConverter(new SigningFailureConverter()),
        Collections.singletonList(new SigningCodec()));
  }

  public static String signatureOf(Payload payload) {
    ByteString signature = payload.getMetadataMap().get(METADATA_KEY);
    return signature == null ? "" : signature.toStringUtf8();
  }

  public static String firstSignature(Payloads payloads) {
    return signatureOf(payloads.getPayloads(0));
  }

  public static HistoryEvent findEvent(
      History history, String name, Predicate<HistoryEvent> predicate) {
    return history.getEventsList().stream()
        .filter(predicate)
        .findFirst()
        .orElseThrow(() -> new AssertionError("no " + name + " event in history"));
  }
}

File: features/serialization_context/sercontext/sercontext.go
// Package sercontext provides a payload codec and failure converter that are
// aware of the serialization context the SDK hands them, plus history helpers
// used by the serialization_context features to assert on the recorded context.
package sercontext

import (
	"context"
	"fmt"

	commonpb "go.temporal.io/api/common/v1"
	enumspb "go.temporal.io/api/enums/v1"
	failurepb "go.temporal.io/api/failure/v1"
	historypb "go.temporal.io/api/history/v1"
	"go.temporal.io/sdk/client"
	"go.temporal.io/sdk/converter"
	"go.temporal.io/sdk/temporal"
	"google.golang.org/protobuf/proto"
)

// MetadataKey is the payload metadata entry Codec stamps with the signature of
// the serialization context it was created with.
const MetadataKey = "ctx-signature"

// NoContext is the signature used when a converter is used without any
// serialization context.
const NoContext = "none"

// DefaultFailureSource is what the default failure converter puts in
// Failure.Source. FailureConverter only overwrites its own fresh failures.
const DefaultFailureSource = "GoSDK"

func WorkflowSignature(namespace, workflowID string) string {
	return fmt.Sprintf("wf|%s|%s", namespace, workflowID)
}

func ActivitySignature(namespace, workflowID, workflowType, activityType, taskQueue string, isLocal bool) string {
	return fmt.Sprintf("act|%s|%s|%s|%s|%s|%t",
		namespace, workflowID, workflowType, activityType, taskQueue, isLocal)
}

// Signature renders a serialization context as a comparable string.
func Signature(ctx converter.SerializationContext) string {
	switch sc := ctx.(type) {
	case converter.WorkflowSerializationContext:
		return WorkflowSignature(sc.Namespace, sc.WorkflowID)
	case converter.ActivitySerializationContext:
		return ActivitySignature(sc.Namespace, sc.WorkflowID, sc.WorkflowType, sc.ActivityType, sc.TaskQueue, sc.IsLocal)
	}
	return NoContext
}

// ClientOptions returns client options wired with the context aware converters.
func ClientOptions() client.Options {
	return client.Options{
		DataConverter:    converter.NewCodecDataConverter(converter.GetDefaultDataConverter(), NewCodec()),
		FailureConverter: NewFailureConverter(),
	}
}

// Codec stamps every payload it encodes with the signature of its serialization
// context and rejects any payload that was encoded under a different one.
type Codec struct {
	signature string
}

func NewCodec() *Codec { return &Codec{signature: NoContext} }

func (c *Codec) WithSerializationContext(ctx converter.SerializationContext) converter.PayloadCodec {
	return &Codec{signature: Signature(ctx)}
}

func (c *Codec) Encode(payloads []*commonpb.Payload) ([]*commonpb.Payload, error) {
	result := make([]*commonpb.Payload, len(payloads))
	for i, p := range payloads {
		clone := proto.Clone(p).(*commonpb.Payload)
		if clone.Metadata == nil {
			clone.Metadata = map[string][]byte{}
		}
		clone.Metadata[MetadataKey] = []byte(c.signature)
		result[i] = clone
	}
	return result, nil
}

func (c *Codec) Decode(payloads []*commonpb.Payload) ([]*commonpb.Payload, error) {
	result := make([]*commonpb.Payload, len(payloads))
	for i, p := range payloads {
		if encoded := SignatureOf(p); encoded != c.signature {
			return nil, fmt.Errorf(
				"serialization context mismatch: payload encoded as %q, decoded as %q", encoded, c.signature)
		}
		clone := proto.Clone(p).(*commonpb.Payload)
		delete(clone.Metadata, MetadataKey)
		result[i] = clone
	}
	return result, nil
}

// FailureConverter records the signature of its serialization context in
// Failure.Source of the failures it creates.
type FailureConverter struct {
	parent    converter.FailureConverter
	signature string
}

func NewFailureConverter() *FailureConverter {
	return &FailureConverter{parent: temporal.GetDefaultFailureConverter(), signature: NoContext}
}

func (f *FailureConverter) WithSerializationContext(ctx converter.SerializationContext) converter.FailureConverter {
	return &FailureConverter{parent: f.parent, signature: Signature(ctx)}
}

func (f *FailureConverter) ErrorToFailure(err error) *failurepb.Failure {
	failure := f.parent.ErrorToFailure(err)
	// A failure that already travelled the wire is returned as-is by the default
	// converter, and its source must not be overwritten.
	if failure != nil && failure.Source == DefaultFailureSource {
		failure.Source = f.signature
	}
	return failure
}

func (f *FailureConverter) FailureToError(failure *failurepb.Failure) error {
	return f.parent.FailureToError(failure)
}

// SignatureOf returns the context signature a payload was encoded with.
func SignatureOf(payload *commonpb.Payload) string {
	return string(payload.GetMetadata()[MetadataKey])
}

// FirstSignature returns the context signature of the first payload.
func FirstSignature(payloads *commonpb.Payloads) string {
	if len(payloads.GetPayloads()) == 0 {
		return ""
	}
	return SignatureOf(payloads.GetPayloads()[0])
}

// Events collects the full history of a workflow execution.
func Events(ctx context.Context, c client.Client, workflowID, runID string) ([]*historypb.HistoryEvent, error) {
	iter := c.GetWorkflowHistory(ctx, workflowID, runID, false, enumspb.HISTORY_EVENT_FILTER_TYPE_ALL_EVENT)
	var events []*historypb.HistoryEvent
	for iter.HasNext() {
		event, err := iter.Next()
		if err != nil {
			return nil, err
		}
		events = append(events, event)
	}
	return events, nil
}

// FindEvent returns the first event matching cond, or an error if there is none.
func FindEvent(
	events []*historypb.HistoryEvent,
	name string,
	cond func(*historypb.HistoryEvent) bool,
) (*historypb.HistoryEvent, error) {
	for _, event := range events {
		if cond(event) {
			return event, nil
		}
	}
	return nil, fmt.Errorf("no %v event in history", name)
}
